chiselFrontend/src/main/scala/chisel3/core/Assert.scala

```diff
@@ -51,12 +51,15 @@ object assert { // scalastyle:ignore object.name
   }
 
   def apply_impl_do(cond: Bool, line: String, message: Option[String], data: Bits*)(implicit sourceInfo: SourceInfo) {
+    val escLine = line.replaceAll("%", "%%")
     when (!(cond || Builder.forcedReset)) {
       val fmt = message match {
-        case Some(str) => s"Assertion failed: $str\n    at $line\n"
-        case None => s"Assertion failed\n    at $line\n"
+        case Some(str) =>
+          val msg = if (data.isEmpty) str.replaceAll("%", "%%") else str
```

Comment:
Right so the basic problem is that people want to be able to write assertions with format strings and arguments and yet also with just normal strings. We **could** require that assert messages are format strings ([there's already a helper to do the error checking](https://github.com/ucb-bar/chisel3/blob/master/chiselFrontend/src/main/scala/chisel3/core/Printable.scala#L37)). My main issue with this is that it strikes me as surprising behavior. Assertions with just simple messages are by far the most common case. However, I agree that it is undesirable to have the same function have entirely different functionality based on if there are any `Data` elements.

One idea is to add another `apply` method that accepts just `Bool` and `String` and another that accepts `Bool`, `String`, `Data`, and `Data*` to differentiate the two forms.